scripts/cmake/utilities.cmake

```diff
@@ -19,3 +22,437 @@ function(mz_target_handle_warnings MZ_TARGET)
+    if(MZ_ADD_NEW_MODULE_TEST_SOURCES)
+        add_custom_target(${MZ_ADD_NEW_MODULE_TARGET_NAME}-alltests)
+        set_target_properties(${MZ_ADD_NEW_MODULE_TARGET_NAME}-alltests PROPERTIES
+            EXCLUDE_FROM_ALL TRUE
+        )
+
+        add_dependencies(build_tests ${MZ_ADD_NEW_MODULE_TARGET_NAME}-alltests)
+
+        set(CPP_TEST_FILES ${MZ_ADD_NEW_MODULE_TEST_SOURCES})
+        list(FILTER CPP_TEST_FILES INCLUDE REGEX "(.*)\\c\\p\\p$")
+        set(QRC_TEST_FILES ${MZ_ADD_NEW_MODULE_TEST_SOURCES})
+        list(FILTER QRC_TEST_FILES INCLUDE REGEX "(.*)\\q\\r\\c$")
+
+        foreach(TEST_FILE ${CPP_TEST_FILES})
+            # The test executable name will be the name of the test file
+            # + the name of the parent target as a prefix.
+            get_filename_component(TEST_NAME ${TEST_FILE} NAME_WE)
+            set(TEST_TARGET_NAME "${MZ_ADD_NEW_MODULE_TARGET_NAME}-${TEST_NAME}")
+
+            mz_add_test_target(
+                TARGET_NAME
+                    ${TEST_TARGET_NAME}
+                TEST_COMMAND
+                    ${TEST_TARGET_NAME}
+                PARENT_TARGET
+                    ${MZ_ADD_NEW_MODULE_TARGET_NAME}
+                SOURCES
+                    ${TEST_FILE}
+                    ${QRC_TEST_FILES}
+                    ${ALL_SOURCES}
+                DEPENDENCIES
+                    ${TEST_LINK_LIBRARIES}
+            )
+
+            # Check if the corresponding header file exists
+            string(REGEX REPLACE ".cpp$" ".h" HEADER_FILE ${TEST_FILE})
+            if(EXISTS ${HEADER_FILE})
+                # Add the header file to the executable if it exists.
+                target_sources(${TEST_TARGET_NAME} PRIVATE ${TEST_FILE})
+            endif()
+        endforeach()
+    endif()
+endfunction()
+
+function(mz_add_library)
+    cmake_parse_arguments(
+        MZ_ADD_LIBRARY # prefix
+        "" # options
+        "" # single-value args
+        "NAME;TYPE" # multi-value args
+        ${ARGN})
+
+    add_library(${MZ_ADD_LIBRARY_NAME} ${MZ_ADD_LIBRARY_TYPE})
+    mz_target_handle_warnings(${MZ_ADD_LIBRARY_NAME})
+    target_compile_definitions(${MZ_ADD_LIBRARY_NAME} PUBLIC
+        "MZ_$<UPPER_CASE:${MZ_PLATFORM_NAME}>"
+        "$<$<CONFIG:Debug>:MZ_DEBUG>"
+    )
+    target_include_directories(${MZ_ADD_LIBRARY_NAME} PRIVATE
+        ${CMAKE_SOURCE_DIR}/src
+        ${CMAKE_CURRENT_SOURCE_DIR}
+    )
+endfunction()
+
+function(mz_add_test_target)
+    cmake_parse_arguments(
+        MZ_ADD_TEST # prefix
+        "" # options
+        "" # single-value args
+        "TARGET_NAME;TEST_COMMAND;PARENT_TARGET;SOURCES;DEPENDENCIES" # multi-value args
+        ${ARGN})
+
+    # Test targets are executable targets.
+    qt_add_executable(${MZ_ADD_TEST_TARGET_NAME}
+        ${MZ_ADD_TEST_SOURCES}
+    )
+    set_target_properties(${MZ_ADD_TEST_TARGET_NAME} PROPERTIES
+        EXCLUDE_FROM_ALL TRUE
```

Comment:
I'm not sure we should keep doing that :) 
This makes ctest wieerd, given that we will be registering the test in ctest, even if it does not exist and there is nothing that will build it. 

In "old" cmake land we have the build_tests target as a proxy to trigger all build things. 
We should keep that, so that in CI we can easy do `cmake build_tests && ctest /build` but also we should have  `-DBUILD_TESTS=OFF` like qt and others that change if tests are added to `ALL` - as that allows for a nice devex i.e test on save.

Reply:
I agree, the use of `EXCLUDE_FROM_ALL` was done simply to try and keep the CI jobs from burning too much CPU time if they weren't going to run tests. But we should probably invert this so that test are built by default and the CI builds can just specify the targets they need (eg: `cmake --build <dir> --target mozillavpn`)